File: Napakalaki JAVA : GUI/src/napakalaki_java/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki_java;
import java.util.*;
import GUI.Dice;

/**
 *
 * @author jumacasni
 */
public class Player {
    static final int MAXLEVEL = 10;
    private String name;
    private int level;
    private boolean dead = true;
    private boolean canISteal = true;
    protected Player enemy;
    private ArrayList<Treasure> visibleTreasures;
    private ArrayList<Treasure> hiddenTreasures;
    private BadConsequence pendingBadConsequence;
    
    public Player(String name){
        this.name = name;
        level = 1;
        enemy = null;
        visibleTreasures = new ArrayList();
        hiddenTreasures = new ArrayList();
        pendingBadConsequence = null;
    }
    
    public Player(Player p){
        name = p.name;
        level = p.level;
        enemy = p.enemy;
        visibleTreasures = p.visibleTreasures;
        hiddenTreasures = p.hiddenTreasures;
        pendingBadConsequence = p.pendingBadConsequence;
    }
    
    public String getName(){
        return name;
    }
    
    private void bringToLife(){
        dead = false;
    }
    
    public int getCombatLevel(){
        int nivel = level;
        for(Treasure tv : visibleTreasures)
            nivel = nivel + tv.getBonus();
        
        return nivel;
    }
            
    private void incrementLevels(int l){
        level = level + l;
    }
    
    private void decrementLevels(int l){
        if(level - l > 0)
            level = level - l;
        else
            level = 1;
    }
    
    private void setPendingBadConsequence(BadConsequence b){
        pendingBadConsequence = b;
    }
    
    private void applyPrize(Monster m){
        int nLevels = m.getLevelsGained();
        this.incrementLevels(nLevels);
        int nTreasures = m.getTreasuresGained();
        
        if(nTreasures > 0){
            CardDealer dealer = CardDealer.getInstance();
            Treasure treasure;
            
            for(int i = 0; i < nTreasures; ++i){
              treasure = dealer.nextTreasure();
              hiddenTreasures.add(treasure);
            }
        }       
    }
    
    private void applyBadConsequence(Monster m){
        BadConsequence badConsequence = m.getBadConsequence();
        int nLevels = badConsequence.getLevels();
        this.decrementLevels(nLevels);
        BadConsequence pendingBad = badConsequence.adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);
        this.setPendingBadConsequence(pendingBad);
    }
    
    private boolean canMakeTreasureVisible(Treasure tesoro){
        boolean sePuede = true;
        

        if(tesoro.getType() == TreasureKind.ONEHAND){
            if(howManyVisibleTreasures(tesoro.getType()) > 1 ||
               howManyVisibleTreasures(TreasureKind.BOTHHANDS) > 0)
                sePuede = false;
        }
        else if(tesoro.getType() == TreasureKind.BOTHHANDS){
            if(howManyVisibleTreasures(tesoro.getType()) > 0 ||
               howManyVisibleTreasures(TreasureKind.ONEHAND) > 0)
                sePuede = false;
        }
        else{
            if(howManyVisibleTreasures(tesoro.getType()) > 0)
                sePuede = false;
        }        
        
        return sePuede;   
    }
    
    private int howManyVisibleTreasures(TreasureKind tKind){
        int nVisibleTreasures = 0;
        for(Treasure t : visibleTreasures)
            if (tKind == t.getType())
                nVisibleTreasures = nVisibleTreasures + 1;
        
        return nVisibleTreasures;
    }
    
    private void dieIfNoTreasures(){
        if(visibleTreasures.isEmpty() && hiddenTreasures.isEmpty())
            dead = true;
    }
    
    public boolean isDead(){
        return dead;
    }
    
    public ArrayList<Treasure> getHiddenTreasures(){
        return hiddenTreasures;
    }
    
    public ArrayList<Treasure> getVisibleTreasures(){
        return visibleTreasures;
    }
    
    public CombatResult combat(Monster m){
        CombatResult combatResult;
        int myLevel = getCombatLevel();
        int monsterLevel = getOponentLevel(m);
        
        if(!canISteal()){
            Dice dice = Dice.getInstance();
            int number = dice.nextNumber();
            if(number < 3){
                int enemyLevel = enemy.getCombatLevel();
                monsterLevel += enemyLevel;
            }
        }
        
        if(myLevel > monsterLevel) {
            applyPrize(m);
            if(level >= MAXLEVEL)
                combatResult = CombatResult.WINGAME;
            else
                combatResult = CombatResult.WIN;
        }
        else{
            applyBadConsequence(m);
            if(shouldConvert())
                combatResult = CombatResult.LOSEANDCONVERT;
            else
                combatResult = CombatResult.LOSE;
        }
            
        return combatResult;
    }
    
    public void makeTreasureVisible(Treasure t){
        boolean canI = canMakeTreasureVisible(t);
        
        if(canI){
            visibleTreasures.add(t);
            hiddenTreasures.remove(t);
        }
    }
    
    public void discardVisibleTreasure(Treasure t){
        visibleTreasures.remove(t);
        if(pendingBadConsequence != null && !pendingBadConsequence.isEmpty())
            pendingBadConsequence.substractVisibleTreasure(t);
        
       dieIfNoTreasures();
    }
    
    public void discardHiddenTreasure(Treasure t){
        hiddenTreasures.remove(t);
        if(pendingBadConsequence != null && !pendingBadConsequence.isEmpty())
            pendingBadConsequence.substractHiddenTreasure(t);
        
        dieIfNoTreasures();       
    }
    
    public boolean validState(){
        
        if(pendingBadConsequence != null){
            if(pendingBadConsequence.isEmpty() && hiddenTreasures.size() <= 4)
                return true;
        }
        else if(hiddenTreasures.size() <= 4)
            return true;
        
        return false;
        
    }
    
    public BadConsequence getPending(){
        return pendingBadConsequence;
    }
            
    
    public void initTreasures(){
        CardDealer dealer = CardDealer.getInstance();
        Dice dice = Dice.getInstance();
        this.bringToLife();
        Treasure treasure = dealer.nextTreasure();
        hiddenTreasures.add(treasure);
        int number = dice.nextNumber();
        if(number > 1){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }
        if(number == 6){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }
    }
    
    public int getLevels(){
        return level;
    }
    
    public Player getEnemy(){
        return enemy;
    }
    
    public Treasure stealTreasure(){
        Treasure treasure = null;
        boolean canI = this.canISteal();
        if(canI){
            boolean canYou = enemy.canYouGiveMeATreasure();
            if(canYou){
                treasure = enemy.giveMeATreasure();
                hiddenTreasures.add(treasure);
                this.haveStolen();
            }
        }
        return treasure;
    }
    
    public void SetEnemy(Player enemy){
        this.enemy = enemy;
    }
    
    protected Treasure giveMeATreasure(){
        int numeroAleatorio;
        numeroAleatorio = (int) (Math.random()*(hiddenTreasures.size()));
        
        Treasure t = hiddenTreasures.get(numeroAleatorio);
        hiddenTreasures.remove(t);
        
        return t;
    }
    
    public boolean canISteal(){
        return canISteal;
    }
    
    protected boolean canYouGiveMeATreasure(){
        if(hiddenTreasures.size() > 0)
            return true;
        
        return false;
    }
    
    private void haveStolen(){
        canISteal = false;
    }
    
    public void discardAllTreasures(){
        ArrayList<Treasure> visibleT = new ArrayList(visibleTreasures);
        ArrayList<Treasure> hiddenT = new ArrayList(hiddenTreasures);
    
        for(Treasure t : visibleT)
            this.discardVisibleTreasure(t);
        
        for(Treasure t : hiddenT)
            this.discardHiddenTreasure(t);
        
    }   
    
    protected int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    
    protected boolean shouldConvert(){
        Dice dice = Dice.getInstance();
        int number = dice.nextNumber();
        if(number == 6)
            return true;
        
        return false;
    }
    
    public String toString(){
        String s = "";
        s = s + name + " (Nivel: " + level + " Combat Level: " + getCombatLevel() + " Enemigo: " + enemy.name + ")";
        if(pendingBadConsequence != null)
            s = s + "\nPendingBadConsequence = " + pendingBadConsequence.toString();
            
        return s;
    }
    
    //Consultores
    protected int getLevel(){
        return level;
    }
    
    protected BadConsequence getPendingBad(){
        return pendingBadConsequence;
    }
    
}

File: Napakalaki RUBY/lib/player.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "napakalaki.rb"

module NapakalakiGame
  class Player
  @@MAXLEVEL = 10
  
  def initialize(n,l,e,ht,vt,pbc)
    @name = n
    @level = l
    @dead = true
    @can_i_steal = true
    @enemy = e
    @hidden_treasures = ht
    @visible_treasures = vt 
    @pending_bad_consequence = pbc
  end
  
  def copyPlayer(p)
    @name = p.getName()
    @level = p.getLevels()
    @dead = p.isDead()
    @can_i_steal = p.canISteal()
    @enemy = p.getEnemy()
    @hidden_treasures = p.getHiddenTreasures() 
    @visible_treasures = p.getVisibleTreasures()
    @pending_bad_consequence = p.getPendingBadConsequence()
  end
  
  def self.newPlayer(n)
      new(n,1,nil,Array.new,Array.new,nil)
  end
  
  def self.getMaxLevel()
    return @@MAXLEVEL
  end
  
  private
   
    def bringToLife()
        @dead = false
    end

    def incrementLevels(l)
      @level = @level + l
    end

    def decrementLevels(l)
      if((@level-l)> 0)
        @level = @level - l
      else
        @level = 1
      end
    end

    def setPendingBadConsequence(b)
      @pending_bad_consequence = b
        
    end

    def applyPrize(m)
        nLevels = m.getLevelsGained()
        incrementLevels(nLevels)
        nTreasures = m.getTreasuresGained()
        
        if(nTreasures > 0)
          dealer = CardDealer.instance
          
          for i in 1..nTreasures
            treasure = dealer.nextTreasure()
            @hidden_treasures << treasure
          end
        end
    end

    def applyBadConsequence(m)
        badConsequence = m.getBadConsequence()
        nLevels = badConsequence.getLevels()
        decrementLevels(nLevels)
        pendingBad = badConsequence.adjustToFitTreasureLists(@visible_treasures, @hidden_treasures)
        setPendingBadConsequence(pendingBad)
    
    end

    def canMakeTreasureVisible(t)
      sePuede = true
      
      if(t.getType() == TreasureKind::ONEHAND)
        if(howManyVisibleTreasures(t.getType()) > 1 || howManyVisibleTreasures(TreasureKind::BOTHHANDS) > 0)
          sePuede = false
        end
      elsif (t.getType() == TreasureKind::BOTHHANDS)
        if(howManyVisibleTreasures(t.getType()) > 0 || howManyVisibleTreasures(TreasureKind::ONEHAND) > 0)
          sePuede = false
        end
      elsif (howManyVisibleTreasures(t.getType()) > 0)
        sePuede = false
      end
      
      return sePuede
    end

    def howManyVisibleTreasures(tKind)
      nVisibleTreasures = 0
      
      @visible_treasures.each do |t|
        if(t.getType() == tKind)
          nVisibleTreasures = nVisibleTreasures + 1
        end
      end
      
      return nVisibleTreasures
    end

    def dieIfNoTreasures()
      if(@visible_treasures.empty? && @hidden_treasures.empty?)
        @dead = true
      end
    end

    def haveStolen()
      @can_i_steal = false
    end
  
  protected
      
    def getOponentLevel(m)
      m.getCombatLevel()
    end
    
    def shouldConvert();
      convert = false
      
      dice = Dice.instance
      number = dice.nextNumber()
      
      if(number == 6)
        convert = true
      end
      
      return convert
    end
    
    def getCombatLevel()
      nivel_total = @level
      @visible_treasures.each {|v| 
        nivel_total = nivel_total + v.getBonus()
      }    
   
      return nivel_total
    end
    
  public
    
    def getEnemy()
      @enemy
    end
    
    
    def canYouGiveMeATreasure()
      if(@hidden_treasures.size > 0)
        return true
      end
        
      return false
    end
    
    def giveMeATreasure()
      numeroAleatorio = rand(0..(@hidden_treasures.size))
      t = @hidden_treasures[numeroAleatorio]
      @hidden_treasures.delete(t)
      return t
    end
    
    def getName()
      @name
    end
    
    def isDead()
      @dead
    end
    
    def getHiddenTreasures()
      @hidden_treasures
    end
    
    def getLevels()
      @level
    end
    
    def canISteal()
      @can_i_steal
    end
    
    def getVisibleTreasures()
      @visible_treasures
    end
    
    def getPendingBadConsequence()
      @pending_bad_consequence
    end

    def combat(m)
       myLevel = getCombatLevel()
       monsterLevel = getOponentLevel(m)
       
      if(!canISteal())
         dice = Dice.instance
         number = dice.nextNumber()
         
        if(number < 3)
          enemyLevel = @enemy.getCombatLevel()
          monsterLevel = monsterLevel + enemyLevel
        end
         
      end
     
      if (myLevel > monsterLevel)
        applyPrize(m)
        if(@level >= @@MAXLEVEL)
          combatResult = CombatResult::WINGAME
        else
          combatResult = CombatResult::WIN
        end
        
      else
        applyBadConsequence(m)
        combatResult = CombatResult::LOSE
        if(shouldConvert())
          combatResult = CombatResult::LOSEANDCONVERT
        end
      end
      
      return combatResult  
    end

    def makeTreasureVisible(t)
        canI = canMakeTreasureVisible(t)
        if(canI)
          @visible_treasures << t
          @hidden_treasures.delete(t)
        end
    end

    def discardVisibleTreasure(t)
        @visible_treasures.delete(t)
        if(@pending_bad_consequence != nil && !(@pending_bad_consequence.isEmpty()))
          @pending_bad_consequence.substractVisibleTreasure(t)
        end
        
        dieIfNoTreasures()
    end

    def discardHiddenTreasure(t)
        @hidden_treasures.delete(t)
        if(@pending_bad_consequence != nil && !(@pending_bad_consequence.isEmpty()))
          @pending_bad_consequence.substractHiddenTreasure(t)
        end
        
        dieIfNoTreasures()
    end

    def validState()
      if(@pending_bad_consequence != nil)
        if(@pending_bad_consequence.isEmpty() && @hidden_treasures.size <= 4)
          return true
        end
      elsif (@hidden_treasures.size <= 4)
        return true
      end
        
      return false 
    end

    def initTreasures()
      dealer = CardDealer.instance
      dice = Dice.instance
      bringToLife()
      treasure = dealer.nextTreasure()
      @hidden_treasures << treasure
      number = dice.nextNumber()
      
      if (number > 1)
        treasure = dealer.nextTreasure()
        @hidden_treasures << treasure
      end
      
      if(number == 6)
        treasure = dealer.nextTreasure()
        @hidden_treasures << treasure
      end
    end

    def stealTreasure()
        treasure = nil
        canI = canISteal()
        if(canI)
          canYou = @enemy.canYouGiveMeATreasure()
          if(canYou)
            treasure = @enemy.giveMeATreasure()
            @hidden_treasures << treasure
            haveStolen()
          end
        end
        
        return treasure
    end

    def setEnemy(enemy)
      @enemy = enemy
    end

    def discardAllTreasures()
      vt = Array.new(@visible_treasures)        
      
      vt.each do |t|
        discardVisibleTreasure(t)
      end
      
      ht = Array.new(@hidden_treasures)
      
      ht.each do |t|
        discardHiddenTreasure(t)
      end
      
    end
    
    def to_s
      s = "#{@name} (Nivel: #{@level} Combat Level: " + getCombatLevel().to_s + " Enemigo: " + @enemy.getName() + ")"
      if(@pending_bad_consequence != nil)
        s = s + "\n Pending Bad Consequence = " + @pending_bad_consequence.to_s
      end
      
      return s
    end
end
end


